### ISSUE TITLE
Set the "background class" as ignored one?

### ISSUE THREAD

**Issue by DISAPPEARED13** [MIC-DKFZ/nnUNet#1293]
Hi, there, I tried to modified a trainer from Generic_UNet to DeeplabV3+ in my segmentation issue. When I just change the self.network definition, and some network architecture, something still goes wrong!
Like, the background class is predicted as one of the target class, so I think might be I need to set "ignore_index=0" while the background class is 0.
Where of the code should I edit? like self.num_classes and the classes of network output just fine?
And: could you please tell me that is it bacause the loss setting so my prediction goes wrong? 
Thanks a lot!

**Comment by FabianIsensee** [comment]
Honestly its impossible to tell without knowing your code.
I suggest you use the `nnUNetTrainerV2_noDeepSupervision` trainer as a basis because your architecture probably doesn't implement deep supervision. nnU-Net expects your network to produce a segmentation map with the same shape as the input. You also need to produce as many classes as nnU-Net expects (see Generic_UNet instantiation!). Do not apply softmax in your architecture. That should work then

**Comment by DISAPPEARED13** [comment]
> 

Hi, there. Sorry to leave the issue for long, and big thanks for your reply. I've been solve the problem after taking much time checking the preprocessing function and the `run_iteration` in the trainer and some other people's code. And finally I found that is the misuage of the Softmax at the `Generic_UNet.final_nonlin` variable (As you said! ).
But what if I want to set the background class not to train and calculate the loss without the background's prediction gradient? Should I just reset the `self.num_class` as (n-1) would be fine?
Great thanks again for your reply! :)

**Comment by DISAPPEARED13** [comment]
> Hi, there. Sorry to leave the issue for long, and big thanks for your reply. I've been solve the problem after taking much time checking the preprocessing function and the `run_iteration` in the trainer and some other people's code. And finally I found that is the misuage of the Softmax at the `Generic_UNet.final_nonlin` variable (As you said! ).
> But what if I want to set the background class not to train and calculate the loss without the background's prediction gradient? Should I just reset the `self.num_class` as (n-1) would be fine?
> Great thanks again for your reply! :)

Hi, there. Sorry to leave the issue for long, and big thanks for your reply. I've been solve the problem after taking much time checking the preprocessing function and the run_iteration in the trainer and some other people's code. And finally I found that is the misuage of the Softmax at the Generic_UNet.final_nonlin variable (As you said! ).
But what if I want to set the background class not to train and calculate the loss without the background's prediction gradient? Should I just reset the self.num_class as (n-1) would be fine?
Great thanks again for your reply! :)

**Comment by DISAPPEARED13** [comment]
> Honestly its impossible to tell without knowing your code. I suggest you use the `nnUNetTrainerV2_noDeepSupervision` trainer as a basis because your architecture probably doesn't implement deep supervision. nnU-Net expects your network to produce a segmentation map with the same shape as the input. You also need to produce as many classes as nnU-Net expects (see Generic_UNet instantiation!). Do not apply softmax in your architecture. That should work then


Hi, there. Sorry to leave the issue for long, and big thanks for your reply. I've been solve the problem after taking much time checking the preprocessing function and the run_iteration in the trainer and some other people's code. And finally I found that is the misuage of the Softmax at the Generic_UNet.final_nonlin variable (As you said! ).
But what if I want to set the background class not to train and calculate the loss without the background's prediction gradient? Should I just reset the self.num_class as (n-1) would be fine?
Great thanks again for your reply! :)

**Comment by FabianIsensee** [comment]
Hey, if you want to treat the background class as an ignore label this is something the current version of nnU-net does not support. But fear not for the new one is on the horizon (couple weeks at most!) and it will support that :-)

**Comment by breAchyz** [comment]
Could you tell me how to output (num_class-1) features at the end of a network and use sigmoid as nolinear function rather than softmax in nnunetv2?

**Comment by FabianIsensee** [comment]
Just use hierarchical labels
https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/region_based_training.md

**Comment by rictoo** [comment]
@FabianIsensee was this ever implemented? Making the background class the ignored class? My data uses integer 0 as the "ignore class" (and it shouldn't be one of the predicted classes during inference).
